internal/cluster/certificates.go

```diff
@@ -0,0 +1,408 @@
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			return errors.WithStack(err)
+		}
+		// If a user has a badly formatted secret it will prevent the cluster from working.
+		kp, err := secretToKeyPair(secret)
+		if err != nil {
+			return err
+		}
+		certificate.KeyPair = kp
+	}
+	return nil
+}
+
+// EnsureAllExist ensure that there is some data present for every certificate
+func (c Certificates) EnsureAllExist() error {
+	for _, certificate := range c {
+		if certificate.KeyPair == nil {
+			return ErrMissingCertificate
+		}
+		if len(certificate.KeyPair.Cert) == 0 {
+			return errors.Wrapf(ErrMissingCrt, "for certificate: %s", certificate.Name)
+		}
+		if len(certificate.KeyPair.Key) == 0 {
+			return errors.Wrapf(ErrMissingKey, "for certificate: %s", certificate.Name)
+		}
+	}
+	return nil
+}
+
+// TODO: consider moving a generating function into the Certificate object itself?
+type certGenerator func() (*certs.KeyPair, error)
+
+// Generate will generate any certificates that do not have KeyPair data.
+func (c Certificates) Generate() error {
+	for _, certificate := range c {
+		if certificate.KeyPair == nil {
+			var generator certGenerator
+			switch certificate.Name {
+			case ServiceAccountName:
+				generator = generateServiceAccountKeys
+			default:
+				generator = generateCACert
+			}
+
+			kp, err := generator()
+			if err != nil {
+				return err
+			}
+			certificate.KeyPair = kp
+			certificate.Generated = true
+		}
+	}
+	return nil
+}
+
+// SaveGenerated will save any certificates that have been generated as Kubernetes secrets.
+func (c Certificates) SaveGenerated(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) error {
+	for _, certificate := range c {
+		if !certificate.Generated {
+			continue
+		}
+		secret := certificate.AsSecret(cluster, config)
+		if err := ctrlclient.Create(ctx, secret); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+	return nil
+}
+
+// GetOrCreate is a convenience function that wraps cluster bootstrap certificate behavior.
+func (c Certificates) GetOrCreate(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster, config *bootstrapv1.KubeadmConfig) error {
+	// Get the certificates that exist
+	if err := c.Lookup(ctx, ctrlclient, cluster); err != nil {
+		return err
+	}
+
+	// Create the certificates that don't exist
+	if err := c.Generate(); err != nil {
+		return err
+	}
+
+	// Save any certificates that have been generated
+	if err := c.SaveGenerated(ctx, ctrlclient, cluster, config); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Certificate represents a single certificate CA.
+type Certificate struct {
+	Generated bool
+	Name      string
+	KeyPair   *certs.KeyPair
+}
+
+// SecretName is the expected name of the Kubernetes secret the certificate is saved in.
+func (c *Certificate) SecretName(clusterName string) string {
+	return fmt.Sprintf("%s-%s", clusterName, c.Name)
```

Comment:
What about calling [this](https://github.com/kubernetes-sigs/cluster-api/blob/9bb49677bc1d8cf97cfea09d1cb0dc3b6d8fd230/util/secret/secret.go#L45) instead of duplicating this func?